### PR TITLE
Move brt utils to pearl

### LINKED ISSUE #348: Move BLOB and Index reading/writing logic from `brt` into `pearl` package
Discussion here: https://github.com/qoollo/bob/pull/262

Currently, `brt` implements `IndexReader`, `BlobReader` and `BlobWriter`. This types depend on the actual file structure, that updates on `pearl` side. For ease of support, it is better to transfer these parts to a `pearl`.

Additionally this will give a support for BPlusTree index structure

### PR DESCRIPTION
**Pull request by vovac12**
Closes #348

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,5 @@
...
-members = [
-    "bob",
-    "bob-apps",
-    "bob-backend",
-    "bob-common",
-]
+members = ["bob", "bob-apps", "bob-backend", "bob-common"]
+
+[patch.crates-io]
+pearl = { git = "https://github.com/qoollo/pearl", branch = "move-brt-utils-to-pearl" }
```
Comment (ikopylov) [inline Cargo.toml:5]: Reminder: update this, when Pearl will be released

File: bob-apps/Cargo.toml [hunk 2]
```diff
@@ -44,6 +47,9 @@ version = "1.14"
 default-features = false
 features = []
 
+[dependencies.pearl]
+version = "=0.9.0"
```
Comment (ikopylov) [inline bob-apps/Cargo.toml:51]: One more reminder: update this, when Pearl will be released

File: bob-apps/bin/blob_recovery_tool/command.rs [hunk 3]
```diff
@@ -201,16 +202,27 @@ pub struct ValidateIndexCommand {
...
     delete: bool,
     skip_confirmation: bool,
+    key_size: Option<usize>,
 }
 
 impl ValidateIndexCommand {
     fn run(&self) -> AnyResult<()> {
+        let validate_index_fn = match self.key_size {
+            Some(1) => validate_index::<Key1>,
+            Some(2) => validate_index::<Key2>,
```
Comment (ikopylov) [inline bob-apps/bin/blob_recovery_tool/command.rs:212]: I think the key sizes `1` and `2` are pointless, so I suggest removing their support (2 bytes key allow to store only 65k records, which is too small for any practical uses of Bob)

File: bob-apps/bin/blob_recovery_tool/command.rs [hunk 4]
```diff
@@ -233,6 +245,9 @@ impl ValidateIndexCommand {
     }
 
     fn subcommand<'a, 'b>() -> App<'a, 'b> {
+        lazy_static::lazy_static! {
+                    static ref KEY_SIZE_HELP: String = format!("key size, supported 1, 2, 4, 8, 16, 32. {} used by default", PearlKey::LEN);
```
Comment (ikopylov) [inline bob-apps/bin/blob_recovery_tool/command.rs:249]: Why is `lazy_static` used? The only place of access to this this field is in the next statement, so it looks like this message can be placed directly as an argument of `.help()`

File: bob-apps/bin/blob_recovery_tool/key.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
+use super::prelude::*;
+
+macro_rules! sized_key {
+    ($t:ident, $n:expr) => {
+        #[derive(Clone, Debug, PartialEq, Eq)]
+        pub struct $t(Vec<u8>);
```
Comment (ikopylov) [inline bob-apps/bin/blob_recovery_tool/key.rs:6]: Key trait was updated in `https://github.com/qoollo/bob/pull/397` (`RefKey` was added), so this macro should be updated accordingly

File: bob-apps/bin/blob_recovery_tool/command.rs (hunk 4 above)
Comment (ikopylov) [reply bob-apps/bin/blob_recovery_tool/command.rs]: @vovac12

File: bob-apps/bin/blob_recovery_tool/key.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+fn le_cmp_keys<const N: usize>(x: &[u8], y: &[u8]) -> std::cmp::Ordering {
+    use std::cmp::Ordering;
+    for i in (0..N).rev() {
+        let ord = x[i].cmp(&y[i]);
+        if ord != Ordering::Equal {
+            return ord;
+        }
+    }
+    Ordering::Equal
```
Comment (ikopylov) [inline bob-apps/bin/blob_recovery_tool/key.rs:11]: For the sake of interest, I tried to optimize this function. The following version looks much better in assembler:
```
fn le_cmp_keys<const N: usize>(x: &[u8], y: &[u8]) -> std::cmp::Ordering {
    use std::cmp::Ordering;
    
    if N % std::mem::size_of::<usize>() == 0 {
        let len = N / std::mem::size_of::<usize>();
        for i in (0..len).rev() {
            let x_part: usize = usize::from_le_bytes(x[i * std::mem::size_of::<usize>()..(i + 1) * std::mem::size_of::<usize>()].try_into().unwrap());
            let y_part: usize = usize::from_le_bytes(y[i * std::mem::size_of::<usize>()..(i + 1) * std::mem::size_of::<usize>()].try_into().unwrap());
            let ord = x_part.cmp(&y_part);
            if ord != Ordering::Equal {
                return ord;
            }
        }
    }
    else {
        for i in (0..N).rev() {
            let ord = x[i].cmp(&y[i]);
            if ord != Ordering::Equal {
                return ord;
            }
        }
    }
    Ordering::Equal
}
```

**Comment by ikopylov** [comment]
@vovac12 please, add a line to CHANGELOG.md

File: bob-apps/bin/blob_recovery_tool/command.rs (hunk 4 above)
Comment (vovac12) [reply bob-apps/bin/blob_recovery_tool/command.rs]: Because App require `'static` lifetime

File: bob-apps/bin/blob_recovery_tool/key.rs (hunk 6 above)
Comment (vovac12) [reply bob-apps/bin/blob_recovery_tool/key.rs:33]: Added

File: bob-apps/bin/blob_recovery_tool/command.rs (hunk 4 above)
Comment (ikopylov) [reply bob-apps/bin/blob_recovery_tool/command.rs]: Thanks for the information

**Review by ikopylov — APPROVED** [review]
LGTM

File: bob-apps/bin/blob_recovery_tool/command.rs [hunk 7]
```diff
@@ -233,6 +243,13 @@ impl ValidateIndexCommand {
     }
 
     fn subcommand<'a, 'b>() -> App<'a, 'b> {
+        lazy_static::lazy_static! {
+            static ref KEY_SIZE_HELP: String =
+                        format!(
+                            "key size, supported 1, 2, 4, 8, 16, 32. {} used by default",
```
Comment (ikopylov) [inline bob-apps/bin/blob_recovery_tool/command.rs:249]:
```suggestion
                            "key size, supported 4, 8, 16, 32. {} used by default",
```

**Pull request merged**